x-pack/plugins/cases/server/services/user_actions/builder.ts

```diff
@@ -0,0 +1,398 @@
+    return this.buildCommonUserAction<'description'>({
+      ...args,
+      action: Actions.update,
+      valueKey: 'description',
+      type: ActionTypes.description,
+      value: args.description,
+    });
+  }
+
+  private buildPushedUserAction(args: PushedBuilderArgs): PushedBuilderResponse {
+    return this.buildCommonUserAction<'pushed'>({
+      ...args,
+      action: Actions.push_to_service,
+      valueKey: 'externalService',
+      value: this.extractConnectorIdFromExternalService(args.externalService),
+      type: ActionTypes.pushed,
+      extraReferences: this.createConnectorPushReference(args.externalService.connector_id),
+    });
+  }
+
+  private buildTagsUserAction(args: TagsBuilderArgs): TagsBuilderResponse {
+    return this.buildCommonUserAction<'tags'>({
+      ...args,
+      valueKey: 'tags',
+      value: args.tags,
+      type: ActionTypes.tags,
+    });
+  }
+
+  private buildStatusUserAction(args: StatusBuilderArgs): StatusBuilderResponse {
+    return this.buildCommonUserAction<'status'>({
+      ...args,
+      action: Actions.update,
+      valueKey: 'status',
+      value: args.status,
+      type: ActionTypes.status,
+    });
+  }
+
+  private buildSettingsUserAction(args: SettingsBuilderArgs): SettingsBuilderResponse {
+    return this.buildCommonUserAction<'settings'>({
+      ...args,
+      action: Actions.update,
+      valueKey: 'settings',
+      value: args.settings,
+      type: ActionTypes.settings,
+    });
+  }
+
+  private buildCreateCaseUserAction(args: CreateCaseBuilderArgs): CreateCaseBuilderResponse {
+    const { payload, caseId, subCaseId, owner, user } = args;
+    const connectorWithoutId = this.extractConnectorId(payload.connector);
+    return {
+      attributes: {
+        ...this.getCommonUserActionAttributes({ user, owner }),
+        action: Actions.create,
+        payload: { ...payload, connector: connectorWithoutId, status: CaseStatuses.open },
+        type: ActionTypes.create_case,
+      },
+      references: [
+        ...this.createCaseReferences(caseId, subCaseId),
+        ...this.createConnectorReference(payload.connector.id),
+      ],
+    } as CreateCaseBuilderResponse;
+  }
+
+  private buildDeleteCaseUserAction(args: DeleteCaseBuilderArgs): DeleteCaseBuilderResponse {
+    const { caseId, owner, user, connectorId } = args;
+    return {
+      attributes: {
+        ...this.getCommonUserActionAttributes({ user, owner }),
+        action: Actions.delete,
+        payload: null,
+        type: ActionTypes.delete_case,
+      },
+      references: [
+        ...this.createCaseReferences(caseId),
+        ...this.createConnectorReference(connectorId),
+      ],
+    };
+  }
+
+  public buildUserAction<T extends BuilderFieldTypes>(
+    field: BuilderFieldTypes,
+    args: BuilderArgumentsType<T>
+  ): BuilderResponseType<T> | null {
+    return this.builderMap.get(field)?.call(this, args) ?? null;
```

Comment:
The only time this would return `undefined/null` is when the caller accidentally passed an invalid type right? It'd be great if we could avoid having to check the response of this method everywhere. 🤔 

I wonder if we should throw? I guess ideally we'd just skip creating a user action instead of creating an error...hmm

Actually better idea, we can return `null/undefined` here and then in the service that calls this we can wrap the `unsecuredSavedObjectClient.create` in a function that does the check and does nothing if it is undefined?

That way we have the `if` in a single place. Would that work?

Reply:
We talked about this, the new format solves this problem 👍